java/src/jni/h5util.c

```diff
@@ -644,27 +644,27 @@ h5str_convert(JNIEnv *env, char **in_str, hid_t container, hid_t tid, void *out_
  *
  * Purpose: Object reference -- show the name of the referenced object.
  *
- * Return:  Nothing
+ * Return:  SUCCEED or FAIL
  *-------------------------------------------------------------------------
  */
-void
-h5str_sprint_reference(JNIEnv *env, h5str_t *out_str, hid_t container, void *ref_p)
+int
+h5str_sprint_reference(JNIEnv *env, h5str_t *out_str, void *ref_p)
 {
-    ssize_t          buf_size;
-    char *           ref_name = NULL;
-    const H5R_ref_t *ref_vp   = (H5R_ref_t *)ref_p;
+    ssize_t    buf_size;
+    char *     ref_name = NULL;
+    H5R_ref_t *ref_vp   = (H5R_ref_t *)ref_p;
 
-    UNUSED(container);
+    int ret_value = FAIL;
 
     if (!h5str_append(out_str, " \""))
-        CHECK_JNI_EXCEPTION(ENVONLY, JNI_FALSE);
+        H5_LIBRARY_ERROR(ENVONLY);
```

Comment:
I think all of these should still be CHECK_JNI_EXCEPTION rather than H5_LIBRARY_ERROR, since they aren't directly calling library functions. If h5str_append calls library functions internally, those should end up handled with an H5_LIBARY_ERROR, but other errors that come from this routine are most likely going to be JNI issues.

Reply:
h5str_append does no error handling of its own or throw any exception, so that is why I changed.

Reply:
It's certainly never going to trigger an H5_LIBRARY_ERROR either though. Probably the only useful thing it can do at the moment is just to skip to the error handling section since h5str_append is so simple that it can only return NULL on error. It probably _should_ be returning an "out of memory error" error internally, but it doesn't..

Reply:
true, so choice is goto done or raiseException.

Reply:
Since the function is to add text to a string, I think if it fails we should raise an exception.

Reply:
Yes I agree with that. The fact that the function just returns NULL on out of memory errors and expects the upper layer to handle it is definitely wrong. We could simply have the function return an out of memory on malloc failure and have the upper layer go back to checking with CHECK_JNI_EXCEPTION.

Reply:
How about H5_JNI_FATAL_ERROR? After all something terrible must have happened and the primary result of calling this function is to print the data.

Reply:
H5_JNI_FATAL_ERROR(ENVONLY. "Unable to append string.");

Reply:
The fatal errors should probably be reserved as a last resort only for errors internal to the JNI, like failing to pin, since it will immediately crash the JVM and application. I think running out of memory isn't severe enough to crash everything. Returning "out of memory" internally is good enough with just a CHECK_JNI_EXCEPTION after h5str_append.

Reply:
Unfortunately, that actual memory allocation happens deeper in a subfunction of h5str_append. 
I will just assume an out of memory;
H5_OUT_OF_MEMORY_ERROR

Reply:
That's what I was implying should be done - put the out of memory error in the internal function and leave a check_exception after h5str_append. Always returning out of memory when str append returns NULL is wrong - it returns NULL in other cases too.

Reply:
If it fails to append, it needs to exit whatever the reason.

Reply:
Yes? But that doesn't mean we should be returning an out of memory error. h5str_append can and will return NULL in other cases than just not being able to allocate space.

Reply:
I didn't want to go into that refactor job.
I will should just return H5_JNI_FATAL_ERROR like we do with HDsprintf calls in the same code functions.

Reply:
It would be a one line change in the internal function. Regardless though, why not just leave these to be CHECK_JNI_EXCEPTIONS like they were before? There's no reason to change them to fatal errors. The reason sprintf calls are fatal errors is because if those fail, a serious system issue actually HAS occurred.

Reply:
CHECK_JNI_EXCEPTION doesn't quit processing:
        if (JNI_TRUE == (*envptr)->ExceptionCheck(envptr)) {                                                 \
            if (JNI_TRUE == clearException)                                                                  \
                (*envptr)->ExceptionClear(envptr);                                                           \
            else                                                                                             \
                goto done;                                                                                   \
        }                                                                                                    \
because there is no java error.

Reply:
There would be a java error if the function fails because it fails to allocate memory..

Reply:
NO;
    if (NULL == (new_str = (char *)HDmalloc(new_len)))
        return;

Reply:
both functions involved:
`/*
 * TODO: no error return. malloc can fail.
 */
void
h5str_resize(h5str_t *str, size_t new_len)
{
    char *new_str;

    if (!str || new_len <= 0 || str->max == new_len)
        return;

    if (NULL == (new_str = (char *)HDmalloc(new_len)))
        return;

    if (new_len > str->max) /* increase memory */
        HDstrcpy(new_str, str->s);
    else
        HDstrncpy(new_str, str->s, new_len - 1);

    HDfree(str->s);
    str->s   = new_str;
    str->max = new_len;
} /* end h5str_resize */

/* appends a copy of the string pointed to by cstr to the h5str.
 Return Value:
 the char string point to str->s
 */
char *
h5str_append(h5str_t *str, const char *cstr)
{
    size_t len;

    if (!str)
        return NULL;
    else if (!cstr)
        return str->s;

    len = HDstrlen(str->s) + HDstrlen(cstr);
    while (len >= str->max) /* not enough to hold the new string, double the space */
    {
        h5str_resize(str, str->max * 2);
    }

    return HDstrcat(str->s, cstr);
} /* end h5str_append */
`

Reply:
This is exactly the point I've been trying to make.. You put the "out of memory error" inside h5str_resize when it fails to malloc, leave the check after h5str_append to be a CHECK_JNI_EXCEPTION, and then everything is fine..

Reply:
Let me state this more directly so we can reach a conclusion here.

Returning out of memory here because you want to exit immediately is probably fine, but is wrong in the sense that h5str_append may return NULL for reasons other than not being able to allocate memory. If we want to do this, fine.

Leaving these as CHECK_JNI_EXCEPTION is fine because it doesn't change what was already there, even though what was already there wasn't handling any real errors.

Returning a fatal error is definitely wrong because not being able to allocate memory is not a reason to crash the whole JVM and application.

Ideally, we should update h5str_resize to return an out of memory error so that the CHECK_JNI_EXCEPTION calls that were at this place previously would actually catch something. Whether you want to do that in this PR or not is up to you, but returning a fatal error is just wrong.

Reply:
Fatal error will not crash the JVM anymore then any other error, H5JNIErrorClass is the common class:
jboolean
h5JNIFatalError(JNIEnv *env, const char *functName)
{
    return H5JNIErrorClass(env, functName, "java/lang/InternalError");
} /* end h5JNIFatalError() */

The refactor is out of scope for this, i.e. a lot more work because of missing args and error handling support.
I will create an issue; HDFFV-11220

Reply:
>Fatal error will not crash the JVM anymore then any other error

No, internal errors WILL crash the JVM. They are handled differently and will immediately crash the JVM with a message similar to

#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (...), pid=12444, tid=9692

Reply:
Okay then I need to review all the other instances to determine which exception to throw.

Reply:
I refactored the JNI code a while back to try and avoid using fatal error as much as possible, but JNI function failures and failures of some c system routines are definitely valid reasons to use a fatal error.

I still think adding an out of memory error to h5str_resize should be a one-line change that's covered because every place that uses that function should be checked with a CHECK_JNI_EXCEPTION, but I also wouldn't be surprised if there are some that aren't.

Reply:
Other then the h5util uses, this is the only other instances
h5gImp.c (3 matches)
225: H5_JNI_FATAL_ERROR(ENVONLY, "H5Gget_info: unable to create H5G_info_t object"); 
257: H5_JNI_FATAL_ERROR(ENVONLY, "H5Gget_info_by_name: unable to create H5G_info_t object"); 
296: H5_JNI_FATAL_ERROR(ENVONLY, "H5Gget_info_by_idx: unable to create H5G_info_t object");

Reply:
Those particular ones may be valid because the underlying functions to create the struct objects are just a bunch of JNI functions being called, so if any of them failed, it's probably a fatal error. The argument could be made that the fatal error should probably be moved internal to those functions and then we could have a normal check at the higher level, but that's more of a grey line.

Reply:
Okay I got a different solution, I added an assertion error - I only need to change h5util functions fron fatal to assertion, which is closer to the intent.